### PR TITLE
:art: ActionData, ActionBase, MultiCall 

### PR DESCRIPTION
**Pull request by zekiblue**

### CODE REVIEW

File: src/interfaces/IActionBase.sol [hunk 1]
```diff
@@ -4,13 +4,23 @@
...
+        uint256[] assetAmounts;
+        // Array with the types of the assets.
+        uint256[] assetTypes;
+    }
     /**
      * @notice Calls a series of addresses with arbitrary calldata.
      * @param actionData A bytes object containing two actionAssetData structs, an address array and a bytes array.
      * @return resultData An actionAssetData struct with the balances of this ActionMultiCall address.
      */
+
```
Comment (Thomas-Smets) [inline src/interfaces/IActionBase.sol:24]: remove whitespace

File: src/actions/MultiCall.sol [hunk 2]
```diff
@@ -86,8 +94,8 @@ contract ActionMultiCall is ActionBase, ERC721TokenReceiver {
             }
         }
 
-        // if any assets were minted and are left in this contract, revert
-        require(mintedIds.length == 0 && mintedAssets.length == 0, "AH: leftover NFTs");
+        // If any assets were minted and are left in this contract, revert.
+        if (mintedIds.length > 0 || mintedAssets.length > 0) revert MultiCallErrors.LEFTOVER_NFTS();
```
Comment (Thomas-Smets) [inline src/actions/MultiCall.sol:98]:
```suggestion
        if (mintedIds.length > 0 || mintedAssets.length > 0) revert MultiCallErrors.LeftoverNfts();
```
Follow error convention

File: src/actions/MultiCall.sol [hunk 3]
```diff
@@ -65,13 +73,13 @@ contract ActionMultiCall is ActionBase, ERC721TokenReceiver {
             if (depositData.assetTypes[i] == 0) {
                 depositData.assetAmounts[i] = IERC20(depositData.assets[i]).balanceOf(address(this));
             } else if (depositData.assetTypes[i] == 1) {
-                // if the amount is 0, we minted a new NFT
+                // If the amount is 0, we minted a new NFT
```
Comment (Thomas-Smets) [inline src/actions/MultiCall.sol:76]:
```suggestion
                // If the amount is 0, we minted a new NFT.
```

File: src/actions/MultiCall.sol (hunk 2 above)
Comment (Thomas-Smets) [reply src/actions/MultiCall.sol:98]: Would keep all errors on this contract (exceptional only for this contract), not on Errors.sol

File: src/actions/MultiCall.sol [hunk 4]
```diff
@@ -40,13 +40,21 @@ contract ActionMultiCall is ActionBase, ERC721TokenReceiver {
 
     /**
      * @notice Calls a series of addresses with arbitrary calldata.
-     * @param actionData A bytes object containing three actionAssetData structs, an address array and a bytes array.
-     * @return resultData An actionAssetData struct with the balances of this ActionMultiCall address.
-     * @dev input address is not used in this generic action.
+     * @param actionData A bytes object containing three actionData structs, an address array and a bytes array.
+     * @return resultData An actionData struct with the balances and ids of this contract address of the given depositData.
```
Comment (Thomas-Smets) [inline src/actions/MultiCall.sol:44]: Still not clear to me, to rephrase

File: src/interfaces/IActionBase.sol (hunk 1 above)
Comment (zekiblue) [reply src/interfaces/IActionBase.sol]: forge fmt gives error check the checks

File: src/actions/MultiCall.sol [hunk 5]
```diff
@@ -43,7 +43,10 @@ contract ActionMultiCall is IActionBase, ERC721TokenReceiver {
      * @notice Calls a series of addresses with arbitrary calldata.
      * @param actionData A bytes object containing three actionData structs, an address array and a bytes array.
      * @return resultData An actionData struct with the balances and ids of this contract address of the given depositData.
-     * @dev Input address is not used in this generic action.
+     * @return depositData The modified `IActionBase.ActionData` struct representing the final state of the `depositData` after executing the action.
+     * @dev It is important to note that the `actionData` must be correctly encoded with the expected argument types.
+     * Otherwise, the execution may fail or produce unexpected results.
+     * Input address is not used in this generic action.
```
Comment (Thomas-Smets) [inline src/actions/MultiCall.sol:49]: Which input address? I still do not get what we try to say here?
What does input address even refer to? @j-vp

Comment (j-vp) [reply src/actions/MultiCall.sol:47]: Think it may refer to another array that was previously in the struct, can be removed

**Comment by j-vp** [comment]
if you take out the struct from the library (just plain in the file), you can do 'import IActionData, ActionData' and remove all the 'IActionData.ActionData' to become simply 'ActionData'

**Review by Thomas-Smets — APPROVED** [review]
gtg

**Pull request merged**